src/probnum/filtsmooth/particlefiltsmooth/_particle_filter.py

```diff
@@ -0,0 +1,218 @@
+    return 1.0 / np.sum(categ_rv.probabilities ** 2)
+
+
+def resample_categorical(
+    categ_rv: random_variables.Categorical,
+) -> random_variables.Categorical:
+    """Resample a categorical random variable."""
+    num_particles = len(categ_rv.support)
+    new_support = categ_rv.sample(size=num_particles)
+    new_event_probs = np.ones(categ_rv.probabilities.shape) / len(
+        categ_rv.probabilities
+    )
+    return random_variables.Categorical(
+        support=new_support,
+        probabilities=new_event_probs,
+        random_state=categ_rv.random_state,
+    )
+
+
+class ParticleFilter(BayesFiltSmooth):
+    r"""Particle filter (PF). Also known as sequential Monte Carlo method.
+
+    A PF estimates the posterior distribution of a Markov process given noisy, non-linear observations,
+    with a set of particles.
+
+    Parameters
+    ----------
+    dynamics_model :
+        Prior dynamics. Since the PF is essentially a discrete-time algorithm,
+        the prior must be a discrete model (or at least one with an equivalent discretisation).
+        This transition must support `forward_realization`.
+    measurement_model :
+        Measurement model. Must be a discrete model that supports `forward_realization`.
+    initrv :
+        Initial random variable. Can be any `RandomVariable` object that implements `sample()`.
+    num_particles :
+        Number of particles to use.
+    linearized_measurement_model :
+        Linearized measurement model that is used as an importance density. In principle,
+        any discrete-time model that supports `backward_realization` is applicable.
+        In practice, it will almost always be one out of `DiscreteEKFComponent`, `DiscreteUKFComponent`,
+        or `IteratedDiscreteComponent`. Linear components are also possible, but would most often imply
+        that a particle filter is not required, because the filtering problem can be used much faster
+        with a Kalman filter. The exception to this rule is if the initial random variable is not Gaussian.
+        Optional. Default is None, which implies the bootstrap PF.
+    with_resampling :
+        Whether after each step the effective number of particles shall be checked, and, if too low,
+        the state should be resampled. Optional. Default is `True`.
+    resampling_percentage_threshold :
+        Percentage threshold for resampling. That is, it is the value :math:`p` such that
+        resampling is performed if :math:`N_{\text{eff}} < p \, N_\text{particles}` holds.
+    """
+
+    def __init__(
+        self,
+        dynamics_model: Union[statespace.LTISDE, statespace.DiscreteGaussian],
+        measurement_model: statespace.DiscreteGaussian,
+        initrv: random_variables.RandomVariable,
+        num_particles: IntArgType,
+        linearized_measurement_model: Optional[statespace.DiscreteGaussian] = None,
+        with_resampling: bool = True,
+        resampling_percentage_threshold: FloatArgType = 0.1,
```

Comment:
one thing that is not part of `ParticleFilter`s currently is any sort of random state. I'd say we leave it for now and wait until the sampling vs. transforming base measure stuff is resolved?

Reply:
Or shall I introduce a `random_state=None` here? your call :)

Reply:
The random state is inferred from the initial random variable.